src/micro.c

```diff
@@ -170,88 +241,158 @@ microGroup_AddEndpoint(microGroup *g, microEndpointConfig *cfg)
     return micro_add_endpoint(NULL, g->m, g, cfg, false);
 }
 
-microError *
-microService_Stop(microService *m)
+static microError *
+_attach_service_to_connection(natsConnection *nc, microService *service)
 {
-    microError *err = NULL;
-    microEndpoint *ep = NULL;
-    bool finalize = false;
-    microDoneHandler doneHandler = NULL;
-
-    if (m == NULL)
+    natsStatus s = NATS_OK;
+    if (nc == NULL || service == NULL)
         return micro_ErrorInvalidArg;
 
-    _lock_service(m);
+    natsConn_Lock(nc);
 
-    if (m->stopped)
+    if (natsConn_isClosed(nc) || natsConn_isDraining(nc))
     {
-        _unlock_service(m);
-        return NULL;
+        natsConn_Unlock(nc);
+        return micro_Errorf("can't add service '%s' to a closed or draining connection", service->cfg->Name);
     }
-    ep = m->first_ep;
 
-    for (; ep != NULL; ep = ep->next)
+    // Wrap the connection callbacks before we subscribe to anything.
+    s = natsOptions_setMicroCallbacks(nc->opts, _on_connection_closed, _on_error);
+
+    if (s == NATS_OK)
     {
-        if (err = micro_stop_endpoint(ep), err != NULL)
+        natsMutex_Lock(nc->servicesMu);
+        if (nc->services == NULL)
         {
-            _unlock_service(m);
-            return microError_Wrapf(err, "failed to stop service '%s', stopping endpoint '%s'", m->cfg->Name, ep->name);
+            nc->services = NATS_CALLOC(1, sizeof(microService *));
+            if (nc->services == NULL)
+                s = nats_setDefaultError(NATS_NO_MEMORY);
+        }
+        else
+        {
+            microService **tmp = NATS_REALLOC(nc->services, (nc->numServices + 1) * sizeof(microService *));
+            if (tmp == NULL)
+                s = nats_setDefaultError(NATS_NO_MEMORY);
+            else
+                nc->services = tmp;
         }
     }
 
-    finalize = (m->first_ep == NULL);
-    if (finalize)
+    if (s == NATS_OK)
     {
-        natsLib_stopServiceCallbacks(m);
-        m->stopped = true;
-        doneHandler = m->cfg->DoneHandler;
+        service->refs++; // no lock needed, called from the constructor
+        nc->services[nc->numServices] = service;
+        nc->numServices++;
     }
+    natsMutex_Unlock(nc->servicesMu);
```

Comment:
The lock was acquired only under `if (s == NATS_OK)` after setting the callbacks, but here you unlock unconditionally. You should the `if` statement above (where you bump service's ref count) under the `if` statement where you acquire the service lock. This unlock would fall under that previous `if` statement too.

Reply:
(hit the wrong button before) thanks for catching this!